### ISSUE TITLE
Add Option to Calculate Median AUROC

### ISSUE THREAD

**Issue by amorehead** [PyTorchLightning/metrics#252]
## 🚀 Feature

Allow the user to calculate the median area under the receiver operating characteristic (AUROC).

### Motivation

In a classification setting where each sample can vary widely in size (e.g. graph neural networks with differing input graph sizes), it would be desirable to choose the median AUROC as the final AUROC value for a given epoch.

### Pitch

A simple option called "median" could be added to TorchMetric's existing AUROC function that allows users to specify that they would like the final AUROC value computed to be the median AUROC value computed over the current epoch.

**Comment by maximsch2** [comment]
Do you mind clarifying what you mean by 'median over the epoch'? Do you mean AUROC is computed separately per batch and then all the per-batch and then you want to store them until the end of the epoch and compute the median?

If you are using Lightning, you can do that right now by returning appropriate things from your `<phase>_step` and then computing median in `<phase>_epoch_end` for example.

In terms of metrics, I think we can have a meter similar to AverageMeter that will compute median, but this will likely requiring storing the entire history of per-batch values.

**Comment by amorehead** [comment]
@maximsch2,

Thank you for reminding me of the `<phase>_step` and `<phase>_epoch_end` hooks in Lightning. I should easily be able to implement a solution to calculate the median AUROC of all the validation and testing dataset batches (separately). My only concern is, in the solution I have outlined below, will DDP facilitate aggregation of all the validation AUROC values for each batch when I go to iterate over all the generated scores in `validation_epoch_end`?

`    def validation_step(self, batch, batch_idx):
        """Lightning calls this inside the validation loop."""
        # Make a forward pass through the network for an entire batch of validation graph pairs
        graphs1, graphs2, labels = batch[0], batch[1], torch.flatten(torch.tensor(batch[2][0], device=self.device))
        logits = torch.flatten(self(graphs1, graphs2))

        # Make predictions
        preds = torch.round(torch.sigmoid(logits))
        int_labels = labels.int()

        # Calculate the batch loss
        bce = self.bce(logits, labels)  # Calculate BCE of a single batch
        self.val_acc(preds, int_labels)  # Calculate Accuracy of a single batch
        val_auroc = self.val_auroc(preds, int_labels)  # Calculate AUROC of a single batch
        self.val_auprc(preds, int_labels)  # Calculate AveragePrecision of a batch

        # Log validation step metric(s)
        self.log('val_bce', bce, sync_dist=True)

        return {'loss': bce, 'val_auroc': val_auroc}

    def validation_epoch_end(self, outputs):
        """Lightning calls this at the end of every validation epoch."""
        self.log('val_acc', self.val_acc.compute())  # Log Accuracy of an epoch
        val_aurocs = torch.tensor([output_dict['val_auroc'] for output_dict in outputs])
        self.log('med_val_auroc', torch.median(val_aurocs))  # Log MedAUROC of an epoch
        self.log('val_auprc', self.val_auprc.compute())  # Log AveragePrecision of an epoch
        self.val_acc.reset()
        self.val_auroc.reset()
        self.val_auprc.reset()`

**Comment by maximsch2** [comment]
No, I think you'll need to communicate those yourself, check out `self.gather_all_tensors` method on LightningModule for that. _epoch_end hook will be called on all workers, but only rank 0 will log, so you want to gather results to that place. There will be a memory cost of aggregating all the results on one node, so be mindful of the size of your dataset there. 

If your dataset size makes this impractical, then there are tricks we can play and potentially provide from torchmetrics:
Similar to `AverageMeter` we can have something like `ApproximatePercentileMeter` that would either do some sort of reservoir-sampling or binning based strategy to find the median (or another percentile)

**Comment by amorehead** [comment]
@maximsch2,

Do you possibly know of any examples I can reference for how to properly use self.all_gather (is this the function you are referring to?) for my LightningModules to collect metric values?

**Comment by SkafteNicki** [comment]
@amorehead this should work:
```python
def validation_step(self, batch, batch_idx):
    """Lightning calls this inside the validation loop."""
    # Make a forward pass through the network for an entire batch of validation graph pairs
    graphs1, graphs2, labels = batch[0], batch[1], torch.flatten(torch.tensor(batch[2][0], device=self.device))
    logits = torch.flatten(self(graphs1, graphs2))

    # Make predictions
    preds = torch.round(torch.sigmoid(logits))
    int_labels = labels.int()

    # Calculate the batch loss
    bce = self.bce(logits, labels)  # Calculate BCE of a single batch
    self.val_acc(preds, int_labels)  # Calculate Accuracy of a single batch
    val_auroc = self.val_auroc(preds, int_labels)  # Calculate AUROC of a single batch
    self.val_auprc(preds, int_labels)  # Calculate AveragePrecision of a batch

    # Log validation step metric(s)
    self.log('val_bce', bce, sync_dist=True)

    return {'loss': bce, 'val_auroc': val_auroc}

def validation_epoch_end(self, outputs):
    """Lightning calls this at the end of every validation epoch."""
    # concatenate for the device
    val_aurocs = torch.cat([output_dict['val_auroc'] for output_dict in outputs]), dim=0)
    # concatenate over all devices
    val_aurocs = torch.cat(self.all_gather(val_aurocs), dim=0)

    self.log('val_acc', self.val_acc.compute())  # Log Accuracy of an epoch
    self.log('med_val_auroc', torch.median(val_aurocs))  # Log MedAUROC of an epoch
    self.log('val_auprc', self.val_auprc.compute())  # Log AveragePrecision of an epoch
    self.val_acc.reset()
    self.val_auroc.reset()
    self.val_auprc.reset()
```

**Comment by amorehead** [comment]
@SkafteNicki,

I'll give your suggestion a try to see how it goes. Thank you very much for the quick reply!

**Comment by amorehead** [comment]
@SkafteNicki,

Just to double-check, what should the size (dimensions) of val_aurocs be after executing "val_aurocs = torch.cat([output_dict['val_auroc'] for output_dict in outputs]), dim=0)"?

![image](https://user-images.githubusercontent.com/7051982/119191267-01763a80-ba44-11eb-9c8f-0ee260977312.png)
